libfs/fs.go

```diff
@@ -0,0 +1,273 @@
+// Copyright 2017 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libfs
+
+import (
+	"context"
+	"os"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/keybase/client/go/logger"
+	"github.com/keybase/kbfs/libkbfs"
+	"github.com/pkg/errors"
+	billy "github.com/src-d/go-billy"
+)
+
+// FS is a wrapper around a KBFS subdirectory that implements the
+// billy.Filesystem interface.  It uses forward-slash separated paths.
+// It may return errors wrapped with the `github.com/pkg/errors`
+// package.
+type FS struct {
+	// Yes, storing ctx in a struct is a mortal sin, but the
+	// billy.Filesystem interface doesn't give us a way to accept ctxs
+	// any other way.
+	ctx      context.Context
+	config   libkbfs.Config
+	root     libkbfs.Node
+	log      logger.Logger
+	deferLog logger.Logger
+}
+
+var _ billy.Filesystem = (*FS)(nil)
+
+// NewFS returns a new FS instance, chroot'd to the given TLF and
+// subdir within that TLF.  `subdir` must exist, and point to a
+// directory, before this function is called.
+func NewFS(ctx context.Context, config libkbfs.Config,
+	tlfHandle *libkbfs.TlfHandle, subdir string) (*FS, error) {
+	rootNode, _, err := config.KBFSOps().GetOrCreateRootNode(
+		ctx, tlfHandle, libkbfs.MasterBranch)
+	if err != nil {
+		return nil, err
+	}
+
+	// Look up the subdir's root.
+	n := rootNode
+	var parts []string
+	if len(subdir) > 0 {
+		parts = strings.Split(subdir, "/")
+	}
+	for i, p := range parts {
+		var ei libkbfs.EntryInfo
+		n, ei, err = config.KBFSOps().Lookup(ctx, n, p)
+		if err != nil {
+			return nil, err
+		}
+		if ei.Type != libkbfs.Dir {
+			return nil, errors.Errorf("%s is not a directory",
+				path.Join(parts[:i]...))
+		}
+	}
+
+	log := config.MakeLogger("")
+	log.CDebugf(ctx, "Made new FS for TLF=%s, subdir=%s",
+		tlfHandle.GetCanonicalName(), subdir)
+
+	return &FS{
+		ctx:      ctx,
+		config:   config,
+		root:     n,
+		log:      log,
+		deferLog: log.CloneWithAddedDepth(1),
+	}, nil
+}
+
+func (fs *FS) lookupOrCreateFile(
+	dir libkbfs.Node, filename string, flag int, perm os.FileMode) (
+	libkbfs.Node, libkbfs.EntryInfo, error) {
+	n, ei, err := fs.config.KBFSOps().Lookup(fs.ctx, dir, filename)
+	switch errors.Cause(err).(type) {
+	case libkbfs.NoSuchNameError:
+		// The file doesn't exist yet; create if requested
+		if flag&os.O_CREATE == 0 {
+			return nil, libkbfs.EntryInfo{}, err
+		}
+		fs.log.CDebugf(
+			fs.ctx, "Creating %s since it doesn't exist yet", filename)
+		excl := libkbfs.NoExcl
+		if flag&os.O_EXCL != 0 {
+			excl = libkbfs.WithExcl
+		}
+		isExec := (perm & 0100) != 0
```

Comment:
KBFS only supports executable permissions for the owner.  This is the same way libfuse behaves.